apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -420,9 +420,39 @@ impl ApplyToInternal for WithRange<PathList> {
                     )
                 }
             }
-            PathList::Expr(expr, tail) => expr
-                .apply_to_path(data, vars, input_path)
-                .and_then_collecting_errors(|value| tail.apply_to_path(value, vars, input_path)),
+
+            PathList::Expr(expr_seq, tail) => {
+                // In case none of the expressions in expr_seq successfully
+                // evaluate to a JSON value, we'll want to see all the errors
+                // that could be responsible for that failure. If any expression
+                // succeeds, all_errors will be ignored.
+                let mut all_errors = Vec::new();
+
+                for expr in expr_seq.as_ref() {
+                    let (expr_value, expr_errors) = expr.apply_to_path(data, vars, input_path);
+                    if let Some(value) = expr_value {
+                        // If any expression evaluates to Some(JSON), apply the
+                        // tail to that value and return errors pertaining only
+                        // to this specific expression or its tail, ignoring any
+                        // errors previously added to all_errors.
+                        return tail
+                            .apply_to_path(&value, vars, input_path)
+                            .prepend_errors(expr_errors);
+                    }
+                    // As long as the expressions keep evaluating to None, keep
+                    // accumulating errors in all_errors.
+                    all_errors.extend(expr_errors);
```

Comment:
so if all subexpressions in `$(a, b, c)`, we'd see four errors? (one for each subexpression plus the `No $(...) expression evaluated successfully`)

is there value in the errors for each subexpression? seems like the last error gives you all the information you need, right?

Reply:
I think there's some value in the subexpression errors because they're the ones that will tell you which property specifically was missing, if the expression could fail in multiple ways.

That said, there definitely is a sort of hierarchical structure to the errors, so maybe it doesn't make sense to lump them all into one array, but instead store the subexpression errors as children of the final error somehow?